syntax = "proto3";

package steinlang;

option cc_enable_arenas = true;

import "language.proto";

message TextInterval {
  int64 begin_line = 2;
  int64 end_line = 3;
  int64 begin_pos = 4;
  int64 end_pos = 5;
}

message SyntaxError {
  string error_message = 1;
  TextInterval location = 2;
}

message AddProgramRequest {
  string key = 1;
  string text = 2;
}

message AddProgramResponse {
  repeated SyntaxError syntax_error = 2;
}

message ModifyProgramRequest {
  string key = 1;
  string text = 2;
}

message ModifyProgramResponse {
  repeated SyntaxError syntax_error = 2;
}

message ViewSnapshotRequest {
  string key = 1;
}

message ViewSnapshotResponse {
  EvalContext snapshot = 1;
}

message GetUpdatesRequest {
  string key = 1;
}

message TextUpdate {
  TextInterval old_location = 1;
  string old_text = 2;
  TextInterval new_location = 3;
  string new_text = 4;
}

message ProgramUpdate {
  repeated string pgm_output = 1;
  repeated TextUpdate pgm_change = 2;
}

service Storage {
  rpc AddProgram(AddProgramRequest) returns(AddProgramResponse) {}
  rpc ModifyProgram(ModifyProgramRequest) returns(ModifyProgramResponse) {}
  rpc ViewSnapshot(ViewSnapshotRequest) returns(ViewSnapshotResponse) {}
  rpc GetUpdates(GetUpdatesRequest) returns(stream ProgramUpdate) {}
}
